<!DOCTYPE html>
<html lang="en" dir="ltr">

<head>
  <!--  required meta tags for bootstrap -->
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <title>DNAvisualization.org</title>
  <script>
    // to tell the JS whether developing locally or on AWS
    var route = '{{ route }}';
  </script>
  {% if STAGE == "production" %}
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-128489917-2"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());
    gtag('config', 'UA-128489917-2');
  </script>
  {% endif %}
  <!-- bootstrap css should before all other css -->
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
  <link rel="stylesheet" href="{{url_for('static',  filename='css/home.css')}}">
  <link rel="stylesheet" href="{{url_for('static',  filename='css/floating-labels.css')}}">

  <link rel="apple-touch-icon" sizes="180x180" href="/icons/apple-touch-icon.png?v=m2Log7W7kM">
  <link rel="icon" type="image/png" sizes="32x32" href="/icons/favicon-32x32.png?v=m2Log7W7kM">
  <link rel="icon" type="image/png" sizes="16x16" href="/icons/favicon-16x16.png?v=m2Log7W7kM">
  <link rel="manifest" href="/icons/site.webmanifest?v=m2Log7W7kM">
  <link rel="mask-icon" href="/icons/safari-pinned-tab.svg?v=m2Log7W7kM" color="#be202e">
  <link rel="shortcut icon" href="/icons/favicon.ico?v=m2Log7W7kM">
  <meta name="msapplication-TileColor" content="#ffc40d">
  <meta name="msapplication-config" content="/icons/browserconfig.xml?v=m2Log7W7kM">
  <meta name="theme-color" content="#ffffff">

</head>

<body>
  <main>
    <!-- nav bar -->
    <nav class="navbar navbar-expand-lg navbar-light bg-light rounded sticky-top">
      <a class="navbar-brand" href="/">
        <img style="max-height: 40px;" src="{{url_for('static', filename='images/logo.svg')}}" alt="DNAvisualization.org">
      </a>
      <a class="navbar-brand d-lg-block d-none" href="/">
        <img style="max-height: 20px;" src="{{url_for('static', filename='images/words.svg')}}">
      </a>
      <button class="navbar-toggler collapsed" type="button" data-toggle="collapse" data-target="#navbarsCollapse" aria-controls="navbarsExample09" aria-expanded="false" aria-label="Toggle navigation">
        <span class="navbar-toggler-icon"></span>
      </button>
      <div class="collapse navbar-collapse justify-content-md-left" id="navbarsCollapse">
        <ul class="navbar-nav ml-auto mt-2 mt-lg-0">
          <li class="nav-item">
            <a class="nav-link" href="#about">About</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#instructions">Instructions</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#architecture">Architecture</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#citation">Citation</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#contact">Contact</a>
          </li>
        </ul>
        <a class="btn btn-outline-secondary" href="https://github.com/Benjamin-Lee/DNAvisualization.org/issues/new?assignees=&labels=bug&template=bug_report.md&title=" target="_blank"><i class="fas fa-bug"></i> Report a bug</a>


      </div>
    </nav>
    <div class="container">
      <!-- Show an error if JS is not enabled -->
      <noscript>
        <div class="alert alert-danger" role="alert">
          You need to enable JavaScript to run this app.
        </div>
      </noscript>
      <div class="jumbotron jumbotron-fluid hide-when-plotting" style="background:transparent !important">
        <div class="container">
          <!-- responsive rendering for the jumbotron -->
          <span class="d-none d-sm-block">
            <h1 class="display-4">Turn <img src="{{url_for('static', filename='images/dna.svg')}}"> into <img src=" {{url_for('static', filename='images/viz.svg')}}">.</h1>
          </span>
          <span class=" d-block d-sm-none">
            <h1 class="display-4">Turn <img src="{{url_for('static', filename='images/dna.svg')}}"></h1>
            <h1 class="display-4">into <img src="{{url_for('static', filename='images/viz.svg')}}">.</h1>
          </span>
          <p class="lead">
            DNAvisualization.org turns DNA sequences into gorgeous, interactive
            two-dimensional visualizations to enable their exploration.
          </p>
        </div>
      </div>
      <div class="row">
        <!--  FYI, col should always be a row! -->
        <div class="col" align="center">
          <!--  This is the plot, it is not visible until the user uploads a fasta file -->
          <div id="hg-container" style="width:100%; height:600px;" class="show-when-plotting my-3 center"></div>
        </div>
      </div>
      <div class="row bg-light border rounded" id="controls">
        <div class="col my-2">
          <!-- the viz method -->
          <div class="hide-when-plotting">
            <!-- control the viz method -->
            <div class="row">
              <div class="col text-center">
                <label for="method">First, choose one or more visualization methods:</label>
              </div>
            </div>
            <div class="row">
              <div class="col text-center">
                <div class="btn-group btn-group-toggle" name="method" id="method" data-toggle="buttons">
                  <label class="btn btn-outline-secondary active" data-toggle="tooltip" data-placement="top" title="(Recommended) Shows variations in GC-content and supports non-ATGC bases.">
                    <input type="checkbox" name="method" id="squiggle" autocomplete="off" checked>Squiggle
                  </label>
                  <!-- <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" title="As, Ts, Cs, and Gs as up, down, left and right, respectively. Cyclic. No support for non-ATGC bases.">
                                <input type="radio" name="method" id="gates" autocomplete="off">Gates
                              </label> -->
                  <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Bases plotted along a unit vector. <var>x</var>-coordinates do not map to base position. No support for non-ATGC bases.">
                    <input type="checkbox" name="method" id="yau" autocomplete="off">Yau
                  </label>
                  <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Same as Yau but projected such that all vectors have lengths in the <var>x</var>-axis of 1 (so that <var>x</var>-coordinates are the same as base positions). No support for non-ATGC bases.">
                    <input type="checkbox" name="method" id="yau-bp" autocomplete="off">Yau-BP
                  </label>
                  <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Like tablature, with As, Ts, Cs, and Gs assigned a <var>y</var>-coordinate. No support for non-ATGC bases. Not recommended.">
                    <input type="checkbox" name="method" id="randic" autocomplete="off">Randic
                  </label>
                  <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Same as Randic, except with dinucleotides instead of bases. Not recommended.">
                    <input type="checkbox" name="method" id="qi" autocomplete="off">Qi
                  </label>
                </div>
              </div>
            </div>
          </div>

          <!-- viz controls during plotting -->
          <div class="show-when-plotting pb-2" id="mode">
            <div class="row">
              <div class="col">
                <!-- Legend mode -->
                <div class="row" id="mode">
                  <div class="col text-center">
                    <label for="mode">Legend Mode</label>
                  </div>
                </div>
                <div class="row">
                  <div class="col text-center">
                    <div class="btn-group btn-group-toggle" name="mode" data-toggle="buttons">
                      <label class="btn btn-outline-secondary active" data-toggle="tooltip" data-placement="top" title="(Recommended) Automatically decide between file and sequence mode.">
                        <input type="radio" name="mode" id="auto" autocomplete="off" checked>Auto
                      </label>
                      <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" title="Plot each file in its own color.">
                        <input type="radio" name="mode" id="file" autocomplete="off">File
                      </label>
                      <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Plot each sequence in its own color.">
                        <input type="radio" name="mode" id="sequence" autocomplete="off">Sequence
                      </label>
                    </div>
                  </div>
                </div>
              </div>
              <div class="col" id="plotting_method_col" style="display:none">
                <!-- viz method -->
                <div class="row" id="method_label">
                  <div class="col text-center">
                    <label for="plotting_method">Method</label>
                  </div>
                </div>
                <div class="row">
                  <div class="col text-center">
                    <div class="btn-group btn-group-toggle" name="plotting_method" id="plotting_method" data-toggle="buttons">
                      <label style="display:none" id="squiggle_plotting_control_label" class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" title="(Recommended) Shows variations in GC-content and supports non-ATGC bases.">
                        <input type="radio" name="plotting_method" id="squiggle_plotting_control" value="squiggle" autocomplete="off">Squiggle
                      </label>
                      <!-- <label class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" title="As, Ts, Cs, and Gs as up, down, left and right, respectively. Cyclic. No support for non-ATGC bases.">
                                    <input type="radio" name="method" id="gates" autocomplete="off">Gates
                                  </label> -->
                      <label style="display:none" id="yau_plotting_control_label" class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Bases plotted along a unit vector. <var>x</var>-coordinates do not map to base position. No support for non-ATGC bases.">
                        <input type="radio" name="plotting_method" id="yau_plotting_control" value="yau" autocomplete="off">Yau
                      </label>
                      <label style="display:none" id="yau-bp_plotting_control_label" class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Same as Yau but projected such that all vectors have lengths in the <var>x</var>-axis of 1 (so that <var>x</var>-coordinates are the same as base positions). No support for non-ATGC bases.">
                        <input type="radio" name="plotting_method" id="yau-bp_plotting_control" value="yau-bp" autocomplete="off">Yau-BP
                      </label>
                      <label style="display:none" id="randic_plotting_control_label" class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Like tablature, with As, Ts, Cs, and Gs assigned a <var>y</var>-coordinate. No support for non-ATGC bases. Not recommended.">
                        <input type="radio" name="plotting_method" id="randic_plotting_control" value="randic" autocomplete="off">Randic
                      </label>
                      <label style="display:none" id="qi_plotting_control_label" class="btn btn-outline-secondary" data-toggle="tooltip" data-placement="top" data-html="true" title="Same as Randic, except with dinucleotides instead of bases. Not recommended.">
                        <input type="radio" name="plotting_method" id="qi_plotting_control" value="qi" autocomplete="off">Qi
                      </label>
                    </div>
                  </div>
                </div>
              </div>
              <div class="col">
                <!-- Options -->
                <div class="row" id="mode">
                  <div class="col text-center">
                    <div class="row">
                      <div class="col">
                        <label for="method">Options</label>
                      </div>
                    </div>
                    <div class="row">
                      <div class="col">
                        <div class="btn-group" role="group" aria-label="Basic example">
                          <button type="button" class="btn btn-outline-secondary" id="remove" data-toggle="tooltip" data-placement="top" title="Remove files"><i class="fas fa-trash"></i></button>
                          <button type="button" class="btn btn-outline-secondary" id="changeTitle" onclick="showTitleModal()" data-toggle="tooltip" data-placement="top" title="Change title and subtitle"><i class="fas fa-bars"></i></button>
                          <div class="btn-group " role="group">
                            <button id="export" type="button" class="btn btn-outline-secondary dropdown-toggle no-carat" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                              <i class="fas fa-file-export"></i>
                            </button>
                            <div class="dropdown-menu" aria-labelledby="export">
                              <a class="dropdown-item" href="#" onclick="chart.print()"><i class="fas fa-print"></i> Print chart</a>
                              <div class="dropdown-divider"></div>
                              <a class="dropdown-item" href="#" id="png"><i class="fas fa-file-image"></i> Download PNG image</a>
                              <a class="dropdown-item" href="#" id="jpeg"><i class="fas fa-file-image"></i> Download JPEG image</a>
                              <a class="dropdown-item" href="#" id="pdf"><i class="fas fa-file-pdf"></i> Download PDF document</a>
                              <a class="dropdown-item" href="#" id="svg"><i class="fas fa-file-code"></i> Download SVG vector image</a>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="col">
                <!-- Zoom controls -->
                <div class="row">
                  <div class="col text-center">
                    <label for="method">Zoom <i class="far fa-question-circle" data-toggle="tooltip" data-placement="top" data-html="true" title="Drag on a region to zoom in and click below to zoom out. Loading the data may take a few moments."></i></label>
                  </div>
                </div>
                <div class="row">
                  <div class="col text-center">
                    <div class="btn-group" role="group" aria-label="Basic example">
                      <label class="btn btn-outline-secondary" id="zoom-out" onclick="zoom(0.5)"><i class="fas fa-search-minus"></i></label>
                      <label class="btn btn-outline-secondary" id="reset-zoom" onclick="default_zoom=true;resetChart(false)">Reset</label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>

          <!--the file adding section -->
          <div class="row">
            <div class="col-12 mt-2">
              <div class="row hide-when-plotting">
                <div class="col text-center">
                  <label>Then, paste or upload a FASTA file:</label>
                </div>
              </div>
              <div class="row">
                <div class="col">
                  <div class="input-group">
                    <div class="custom-file mb-2" id="dropzone">
                      <input type="file" class="custom-file-input" id="file-input" multiple accept=".fa,.fasta,.fna,.txt">
                      <label class="custom-file-label"><i class="fas fa-file"></i> <span class="d-none d-sm-inline">Choose FASTA file or drop anywhere on page...</span><span class="d-xs-inline d-sm-none">Choose FASTA file...</span>
                      </label>
                    </div>
                  </div>
                </div>
                <div class="col-md-3">
                  <div class="btn-group btn-block">
                    <button type="button" class="btn btn-outline-secondary btn-block rounded" id="paste-sequence"><i class="fas fa-paste"></i> Paste FASTA </button>
                    <button type="button" hidden class="btn btn-outline-secondary dropdown-toggle dropdown-toggle-split" id="pasted_sequence_toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                      <span class="sr-only">Toggle Dropdown</span>
                    </button>
                    <div class="dropdown-menu" id="pasted_sequence_dropdown">
                    </div>
                  </div>
                </div>
              </div>
              <div class="row hide-when-plotting mt-2">
                <div class="col mb-2">
                  <div class="btn-group btn-block hide-when-plotting">
                    <button type="button" class="btn btn-outline-secondary btn-block" id="load-example"><i class="fas fa-dna"></i> Load an example file</button>
                    <button type="button" class="btn btn-outline-secondary dropdown-toggle dropdown-toggle-split" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                      <span class="sr-only">Toggle Dropdown</span>
                    </button>
                    <div class="dropdown-menu">
                      <a class="dropdown-item" href="#" onclick="plotSequence (hbb, 'hbb.fasta')">Hemoglobin subunit beta, multiple species (1,871 bp)</a>
                      <a class="dropdown-item" href="#" onclick="plotSequence (titin, 'titin.fasta')">Human titin transcript variant N2BA, mRNA (105,223 bp)</a>
                      <a class="dropdown-item" href="#" onclick="plotSequence (bsub, 'GCA_000293765.1_ASM29376v1_genomic.fna')"><em>Bacillus subtilis</em> QB928 complete genome (4,195,853 bp)</a>
                    </div>
                  </div>
                </div>
                <div class="col-lg mb-2">
                  <div class="btn-group btn-block hide-when-plotting">
                    <a href="{{url_for('static',  filename='files/hbb.fna')}}" class="btn btn-outline-secondary btn-block hide-when-plotting" id="download-example"><i class="fas fa-file-download"></i> Download example file</a>
                    <button type="button" class="btn btn-outline-secondary dropdown-toggle dropdown-toggle-split" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                      <span class="sr-only">Toggle Dropdown</span>
                    </button>
                    <div class="dropdown-menu">
                      <a class="dropdown-item" href="{{url_for('static',  filename='files/hbb.fna')}}">Hemoglobin subunit beta, multiple species (1,871 bp)</a>
                      <a class="dropdown-item" href="{{url_for('static',  filename='files/titin.fna')}}">Human titin transcript variant N2BA, mRNA (105,223 bp)</a>
                      <a class="dropdown-item" href="{{url_for('static',  filename='files/GCA_000293765.1_ASM29376v1_genomic.fna')}}"><em>Bacillus subtilis</em> QB928 complete genome (4,195,853 bp)</a>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>




      <!-- the about section -->
      <div class="row pt-3">
        <div class="col">
          <h2 id="about">About</h2>

          <p>Raw DNA sequences, which consist of long strings of letters, are
            hard for humans to make sense of at a glance. However, there is
            significant biological meaning contained within them. To reveal this
            meaning, <a href="http://squiggle.readthedocs.io/en/latest/methods.html">a number
              of methods</a> have been proposed to convert raw DNA sequences into
            two-dimensional visualizations. This website allows you to try several
            of these methods out on your data with a high-performance
            parallel-computing backend enabling genome-scale visualization.</p>

          <h2 id="instructions">Instructions</h2>

          <p>Using this website is easy:</p>

          <ol>
            <li>Choose a visualization method. For full information about the
              visualization methods, take a look <a href="http://squiggle.readthedocs.io/en/latest/methods.html">here</a>.</li>
            <li>Upload a sequence by: <ul>
                <li>Clicking the "browse" button</li>
                <li>Dragging a file anywhere on the page</li>
                <li>Loading an example file</li>
                <li>Pasting a file while anywhere on the page</li>
                <li>Pasting the contents of a FASTA file into a textbox</li>
              </ul>
            </li>
          </ol>

          <p class="font-weight-bold">That's it.</p>

          <p>Optionally, after plotting a sequence, you can:</p>

          <ul>
            <li>Remove files from the visualization by clicking the <i class="fas fa-trash"></i> icon and selecting files</li>
            <li>Modify the graph's title and subtitle by clicking the <i class="fas fa-bars"></i> icon</li>
            <li>Export the content of the graph in .png, .jpeg, .svg, or .pdf format by clicking on the <i class="fas fa-file-export"></i> icon</li>
            <li>Zoom in on a region of interest by clicking and dragging over it, zoom out via the <i class="fas fa-search-minus"></i> icon, or reset the zoom.</li>
          </ul>

          <p>When you plot your data using this website, there are a few things
            to remember. By default, each DNA sequence you wish to display will be
            rendered in a different color unless there are more sequences than
            there are availible colors, in which case each file will be shown in
            its own color. This may be overridden at any time using the "legend
            mode" selector that becomes visible after plotting. Although the
            website can display up to thirty sequences of up to 4.5 Mbp each
            simultaneously, you may find that the resulting graph becomes very
            cluttered.</p>

          <p>On the "visualization method" click-box, you choose how to display
            your data using the three different display methods currently
            supported. We are actively investigating adding more methods.</p>

          <p>As you look at an output graph with several overlaid sequences, you
            will see that when the sequences completely mirror one another, that
            the sequences are essentially identical. When you see the plot lines
            begin to diverge, you will instantly be able to see where the
            sequences have begun to differ. You can zoom in for a closer view,
            then reset the zoom to enable a return the original (unzoomed) view.
            When the sequences are parallel on the plot, that tells you that the
            sequences are once again identical after a period of divergence.</p>

          <p>Try visualizing your data with each of the visualization
            techniques; each method shows your data from a slightly different
            perspective, but be aware of each method’s strengths or weaknesses so
            that you use the right tool for the right job.</p>

          <p>When you finish using the site, you can output your data using the
            .png, .pdf, or .jpg formats in a publication-ready graph. Your data
            will be retained for 24 hours, but after that time has elapsed, you
            will need to re-upload your data if you wish to continue using this
            tool.</p>

          <h2 id="architecture">Architecture</h2>

          <p>DNAvisualization.org is built on an entirelesly serverless
            architecture. Due to the inherently parallelizable nature of DNA
            sequence transformation (the transformation of one sequence has no
            effect on the transformation of another), this design yields a
            significant performance increase at a lower cost compared to a
            traditional architecture.</p>

          <p>This website uses <a href="http://flask.pocoo.org">Flask</a> as a
            lightweight Python web framework and <a href="https://www.zappa.io">Zappa</a> to automate deployment to Amazon
            Web Service's (AWS) <a href="https://aws.amazon.com/lambda">Lambda</a>
            serverless computing platform. Additionally, it takes advantace of <a href="https://aws.amazon.com/s3/"> AWS S3</a> for serverless data
            storage and querying. All stored data are automatically deleted after 24 hours.</p>

          <p>For a full overview of the architecture, take a look at the <a href="https://en.wikipedia.org/wiki/Sequence_diagram">sequence diagram</a> below:</p>

          <div class="mermaid text-center">
            {% include "test.mmd" without context %}
          </div>

          <h2 id="citation" class="pt-2">Citation</h2>

          <p>If you use this website in your research, please cite is as:</p>
          <ul>
            <li>Lee B. D., Timony M. A., & Ruiz P. (2019). <a href="https://DNAvisualization.org">DNAvisualization.org.</a></li>
          </ul>
          <p>Additionally, please be sure to cite the visualization method you
            used. Click below to display the citation information for each method:</p>
          <details>
            <summary>Squiggle</summary>
            <p>If you use the Squiggle visualization method, please cite the
              following paper <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty807/5104938?guestAccessKey=306f0370-1281-428b-aa8a-757d3c49d801">(full
                text)</a>:</p>
            <ul>
              <li> Lee, B. D. (2018). Squiggle: a user-friendly two-dimensional DNA sequence
                visualization tool. <span class="font-italic">Bioinformatics</span>. doi:<a href="https://doi.org/10.1093/bioinformatics/bty807">10.1093/bioinformatics/bty807</a>.</li>
            </ul>
          </details>
          <details>
            <summary>Yau</summary>

            <p>If you use the Yau visualization method, please cite the
              following paper <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC162336/">(full
                text)</a>:</p>
            <ul>
              <li>Yau, S. S., Wang, J., Niknejad, A., Lu, C., Jin, N., & Ho, Y.
                K. (2003). DNA sequence representation without degeneracy. <span class="font-italic">Nucleic acids research, 31</span>(12),
                3078-80.</li>
            </ul>
          </details>
          <details>
            <summary>Yau-BP</summary>
            <p>If you use the Yau-BP visualization method, please cite the
              following papers:</p>
            <ul>
              <li> Lee, B. D. (2018). Squiggle: a user-friendly two-dimensional
                DNA sequence visualization tool. <span class="font-italic">Bioinformatics</span>. doi:<a href="https://doi.org/10.1093/bioinformatics/bty807">10.1093/bioinformatics/bty807</a>.</li>
              <li>Yau, S. S., Wang, J., Niknejad, A., Lu, C., Jin, N., & Ho, Y.
                K. (2003). DNA sequence representation without degeneracy. <span class="font-italic">Nucleic acids research, 31</span>(12),
                3078-80.</li>
            </ul>
          </details>
          <details>
            <summary>Randic</summary>
            <p>If you use the Randic visualization method, please cite the
              following paper:</p>
            <ul>
              <li>Randić, M., Vračko, M., Lerš, N., & Plavšić, D. (2003). Novel 2-D graphical representation of DNA sequences and their numerical characterization. <span class="font-italic">Chemical Physics Letters, 368</span>(1–2), 1–6. doi:<a href="https://doi.org/10.1016/s0009-2614(02)01784-0">10.1016/s0009-2614(02)01784-0</a>.</li>
            </ul>
          </details>
          <details class="pb-3">
            <summary>Qi</summary>
            <p>If you use the Qi visualization method, please cite the
              following paper:</p>
            <ul>
              <li>Qi, Z., & Qi, X. (2007). Novel 2D graphical representation of
                DNA sequence based on dual nucleotides. <span class="font-italic">Chemical Physics Letters, 440</span>(1–3),
                139–144. doi:<a href="https://doi.org/10.1016/j.cplett.2007.03.107">10.1016/j.cplett.2007.03.107</a>.</li>
            </ul>
          </details>
          <h2 id="contact">Contact Us</h2>

          <p>Having issues? We're happy to help get them resolved. If you think you've found a bug, the best way to get in touch is to <a href="https://github.com/Benjamin-Lee/DNAvisualization.org/issues/new?assignees=&labels=bug&template=bug_report.md&title="
              target="_blank"> make an bug report</a>
            on the project's <a href="https://github.com/Benjamin-Lee/DNAvisualization.org">GitHub repository</a>.</p>

          <p>Have an idea for another way to turn a DNA sequence into a
            two-dimensional visualization? Let us know over on the <a href="https://github.com/Lab41/squiggle/issues/new"> Squiggle repository</a> and we'll be happy to work with you on implementing it.
            (DNAvisualization.org uses a stripped-down version of the Squiggle
            library to transform sequences.)
          </p>

          <p>To make a feature request for the DNAvisualization.org website, please click <a href="https://github.com/Benjamin-Lee/DNAvisualization.org/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=" target="_blank">here</a>.
          </p>
        </div>
      </div>
    </div><!--  /container -->
  </main>

  <footer class="footer py-3 bg-light">
    <div class="container text-center">
      <span class="text-muted">&copy;2019 IQT Labs LLC, a wholly owned research venture of <a class="text-muted" href="https://iqt.org">In-Q-Tel, Inc.</a> This content is released under <a class="text-muted" target="_blank" href="https://github.com/Benjamin-Lee/DNAvisualization.org/blob/master/LICENSE">
          the Apache 2.0 License</a>, except for the DNA and chart icons, which are licensed under <a href="https://creativecommons.org/licenses/by-nd/3.0/" class="text-muted">CC BY-ND 3.0</a> by <a class="text-muted" href="https://icons8.com">Icons8</a>.</span>
    </div>
  </footer>

  <!-- placed at bottom to speed page loading -->
  <!-- polyfill first -->
  <script src="https://cdn.polyfill.io/v2/polyfill.min.js"></script>

  <!-- jQuery first, then Popper.js, then Bootstrap JS -->
  <!--  bootstrap requirements -->
  <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
  <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>

  <!-- Highcharts -->
  <script src="https://code.highcharts.com/highcharts.js"></script>
  <script src="https://code.highcharts.com/modules/boost.js"></script>
  <script src="https://code.highcharts.com/modules/exporting.js"></script>

  <!-- other utilities -->
  <script src="https://unpkg.com/axios/dist/axios.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/lodash@4.17.11/lodash.min.js" integrity="sha256-7/yoZS3548fXSRXqc/xYzjsmuW3sFKzuvOCHd06Pmps=" crossorigin="anonymous"></script>
  <script defer src="https://use.fontawesome.com/releases/v5.6.3/js/all.js" integrity="sha384-EIHISlAOj4zgYieurP0SdoiBYfGJKkgWedPHH4jCzpCXLmzVsw1ouK59MuUtP4a1" crossorigin="anonymous"></script>
  <script src="https://unpkg.com/mermaid@8.0.0/dist/mermaid.min.js"></script>
  </script>
  <!-- locally hosted scripts -->
  <script type="text/javascript" src="{{ url_for('static', filename='js/filereader.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/xxhash.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/reformat.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/bootbox.all.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/dragster.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/modernizr.js') }}"></script>


  <!-- last, load the JS for the page -->
  <script type="text/javascript" src="{{ url_for('static', filename='js/home.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/example_seqs.js') }}"></script>

</body>

</html>
